sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -119,9 +119,21 @@ object NestedColumnAliasing {
       .filter(!_.references.subsetOf(AttributeSet(otherRootReferences)))
       .groupBy(_.references.head)
       .flatMap { case (attr, nestedFields: Seq[ExtractValue]) =>
+        // Remove redundant `ExtractValue`s if they share the same parent nest field.
+        // For example, when `a.b` and `a.b.c` are in project list, we only need to alias `a.b`.
+        // We only need to deal with two `ExtractValue`: `GetArrayStructFields` and
+        // `GetStructField`. Please refer to the method `collectRootReferenceAndExtractValue`.
+        val dedupNestedFields = nestedFields.filter {
+          case a @ GetArrayStructFields(child, _, _, _, _) =>
+            nestedFields.forall(f => f == a || child.find(_.semanticEquals(f)).isEmpty)
+          case n @ GetStructField(child, _, _) =>
+            nestedFields.forall(f => f == n || child.find(_.semanticEquals(f)).isEmpty)
```

Comment:
Can we merge the two entry into one? e.g.,
```
          case e @ (_: GetStructField | _: GetArrayStructFields) =>
            nestedFields.forall(f => f == e || e.children.head.find(_.semanticEquals(f)).isEmpty)
```

Reply:
Also, it seems the added test below passes without `f == n`? If so, can you add a test for it?

Reply:
@maropu Thanks. Will address two points.

Reply:
I think f == a condition is not necessary now. It should be left here by previous commit.

Reply:
Thanks for the check! Ur, one more comment: can we avoid the loop of `length(nestedFields)^2`? in: https://github.com/apache/spark/pull/24525/files#diff-43334bab9616cc53e8797b9afa9fc7aaR129

Reply:
Sorry for reply late.

I thought about this. We can check for a nested field accessor in other accessors which have less ExtractValue. For example, we don't need to check if `a.b` is redundant by looking at `a.b.c`, `a.b.c.d`, etc.

However to do this, we first need to collect the info about number of ExtractValue. Is it worth?

Reply:
Ur, I see. Yea, I think its ok as it is. thanks for the check.